File: planner/memo/expr_iterator.go
// Copyright 2018 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// See the License for the specific language governing permissions and
// limitations under the License.

package memo

import (
	"container/list"
)

// ExprIter enumerates all the equivalent expressions in the Group according to
// the expression pattern.
type ExprIter struct {
	// Group and Element solely identify a Group expression.
	*Group
	*list.Element

	// matched indicates whether the current Group expression bound by the
	// iterator matches the pattern after the creation or iteration.
	matched bool

	// Pattern describes the node of pattern tree.
	// The Operand type of the Group expression and the EngineType of the Group
	// must be matched with it.
	*Pattern

	// Children is used to iterate the child expressions.
	Children []*ExprIter
}

// Next returns the next Group expression matches the pattern.
func (iter *ExprIter) Next() (found bool) {
	defer func() {
		iter.matched = found
	}()

	// Iterate child firstly.
	for i := len(iter.Children) - 1; i >= 0; i-- {
		if !iter.Children[i].Next() {
			continue
		}

		for j := i + 1; j < len(iter.Children); j++ {
			iter.Children[j].Reset()
		}
		return true
	}

	// It's root node or leaf ANY node.
	if iter.Group == nil || iter.Operand == OperandAny {
		return false
	}

	// Otherwise, iterate itself to find more matched equivalent expressions.
	for elem := iter.Element.Next(); elem != nil; elem = elem.Next() {
		expr := elem.Value.(*GroupExpr)

		if !iter.Operand.Match(GetOperand(expr.ExprNode)) {
			// All the Equivalents which have the same Operand are continuously
			// stored in the list. Once the current equivalent can not Match
			// the Operand, the rest can not, either.
			return false
		}

		if len(iter.Children) == 0 {
			iter.Element = elem
			return true
		}
		if len(iter.Children) != len(expr.Children) {
			continue
		}

		allMatched := true
		for i := range iter.Children {
			iter.Children[i].Group = expr.Children[i]
			if !iter.Children[i].Reset() {
				allMatched = false
				break
			}
		}

		if allMatched {
			iter.Element = elem
			return true
		}
	}
	return false
}

// Matched returns whether the iterator founds a Group expression matches the
// pattern.
func (iter *ExprIter) Matched() bool {
	return iter.matched
}

// Reset resets the iterator to the first matched Group expression.
func (iter *ExprIter) Reset() (findMatch bool) {
	defer func() { iter.matched = findMatch }()

	if iter.Pattern.MatchOperandAny(iter.Group.EngineType) {
		return true
	}

	for elem := iter.Group.GetFirstElem(iter.Operand); elem != nil; elem = elem.Next() {
		expr := elem.Value.(*GroupExpr)

		if !iter.Pattern.Match(GetOperand(expr.ExprNode), expr.Group.EngineType) {
			break
		}

		// The leaf node of the pattern tree might not be an OperandAny or a XXXScan.
		// We allow the patterns like: Selection -> Projection.
		// For example, we have such a memo:
		// Group#1
		//     Selection_0 input:[Group#2]
		// Group#2
		//     Projection_1 input:[Group#3]
		//     Projection_2 input:[Group#4]
		// Group#3
		//     .....
		// For the pattern above, we will match it twice: `Selection_0->Projection_1`
		// and `Selection_0->Projection_2`. So if the iterator has no children, we can safely return
		// the element here.
		if len(iter.Children) == 0 {
			iter.Element = elem
			return true
		}
		if len(expr.Children) != len(iter.Children) {
			continue
		}

		allMatched := true
		for i := range iter.Children {
			iter.Children[i].Group = expr.Children[i]
			if !iter.Children[i].Reset() {
				allMatched = false
				break
			}
		}
		if allMatched {
			iter.Element = elem
			return true
		}
	}
	return false
}

// GetExpr returns the root GroupExpr of the iterator.
func (iter *ExprIter) GetExpr() *GroupExpr {
	return iter.Element.Value.(*GroupExpr)
}

// NewExprIterFromGroupElem creates the iterator on the Group Element.
func NewExprIterFromGroupElem(elem *list.Element, p *Pattern) *ExprIter {
	expr := elem.Value.(*GroupExpr)
	if !p.Match(GetOperand(expr.ExprNode), expr.Group.EngineType) {
		return nil
	}
	iter := newExprIterFromGroupExpr(expr, p)
	if iter != nil {
		iter.Element = elem
	}
	return iter
}

// newExprIterFromGroupExpr creates the iterator on the Group expression.
func newExprIterFromGroupExpr(expr *GroupExpr, p *Pattern) *ExprIter {
	if len(p.Children) != 0 && len(p.Children) != len(expr.Children) {
		return nil
	}
	iter := &ExprIter{Pattern: p, matched: true}
	for i := range p.Children {
		childIter := newExprIterFromGroup(expr.Children[i], p.Children[i])
		if childIter == nil {
			return nil
		}
		iter.Children = append(iter.Children, childIter)
	}
	return iter
}

// newExprIterFromGroup creates the iterator on the Group.
func newExprIterFromGroup(g *Group, p *Pattern) *ExprIter {
	if p.MatchOperandAny(g.EngineType) {
		return &ExprIter{Group: g, Pattern: p, matched: true}
	}
	for elem := g.GetFirstElem(p.Operand); elem != nil; elem = elem.Next() {
		expr := elem.Value.(*GroupExpr)
		if !p.Match(GetOperand(expr.ExprNode), g.EngineType) {
			return nil
		}
		iter := newExprIterFromGroupExpr(expr, p)
		if iter != nil {
			iter.Group, iter.Element = g, elem
			return iter
		}
	}
	return nil
}

File: rust/difftests/planner-tests/tests/expr_iterator.rs
// Copyright 2026 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Dependency-closed vectors for `pkg/planner/memo/expr_iterator.go`.
//!
//! The direct Go anchors are `TestNewExprIterFromGroupElem` at line 29,
//! `TestExprIterNext` at line 77, `TestExprIterReset` at line 130, and
//! `TestExprIterWithEngineType` at line 207 in
//! `pkg/planner/memo/expr_iterator_test.go`.

use tidb_planner::expr_iterator::{
    new_expr_iter_from_group, new_expr_iter_from_group_elem, Group, GroupExpression,
};
use tidb_planner::pattern::{new_pattern, Operand};
use tidb_planner::pattern_engine::{EngineType, EngineTypeSet};

fn group(engine: EngineType, operands: &[Operand]) -> Group {
    let mut value = Group::new(engine);
    for &operand in operands {
        value.insert(GroupExpression::new(operand));
    }
    value
}

fn join_group(left: Group, right: Group) -> Group {
    let mut root = Group::new(EngineType::TiDb);
    root.insert(
        GroupExpression::new(Operand::Join)
            .with_child(left)
            .with_child(right),
    );
    root
}

#[test]
fn source_group_element_constructor_matches_root_and_children() {
    let root = join_group(
        group(
            EngineType::TiDb,
            &[Operand::Selection, Operand::Projection, Operand::Limit],
        ),
        group(
            EngineType::TiDb,
            &[Operand::Limit, Operand::Selection, Operand::Projection],
        ),
    );
    let pattern = new_pattern(Operand::Join, EngineTypeSet::ALL);
    let iter = new_expr_iter_from_group_elem(&root, 0, &pattern).expect("root match");
    assert!(iter.matched());
    assert_eq!(iter.current().expect("current").operand, Operand::Join);
    assert_eq!(iter.current().expect("current").children.len(), 0);
}

#[test]
fn source_next_enumerates_cartesian_child_matches() {
    let root = join_group(
        group(
            EngineType::TiDb,
            &[Operand::Projection, Operand::Limit, Operand::Projection],
        ),
        group(
            EngineType::TiDb,
            &[Operand::Selection, Operand::Limit, Operand::Selection],
        ),
    );
    let pattern = tidb_planner::pattern::build_pattern(
        Operand::Join,
        EngineTypeSet::ALL,
        [
            new_pattern(Operand::Projection, EngineTypeSet::ALL),
            new_pattern(Operand::Selection, EngineTypeSet::ALL),
        ],
    );
    let mut iter = new_expr_iter_from_group(&root, &pattern).expect("join matches");
    assert_eq!(iter.len(), 4);
    let mut count = 0;
    while iter.matched() {
        let current = iter.current().expect("current match");
        assert_eq!(current.children.len(), 2);
        count += 1;
        iter.advance();
    }
    assert_eq!(count, 4);
    assert!(!iter.matched());
}

#[test]
fn source_reset_returns_to_first_match() {
    let root = join_group(
        group(
            EngineType::TiDb,
            &[Operand::Projection, Operand::Projection],
        ),
        group(EngineType::TiDb, &[Operand::Selection]),
    );
    let pattern = tidb_planner::pattern::build_pattern(
        Operand::Join,
        EngineTypeSet::ALL,
        [
            new_pattern(Operand::Projection, EngineTypeSet::ALL),
            new_pattern(Operand::Selection, EngineTypeSet::ALL),
        ],
    );
    let mut iter = new_expr_iter_from_group(&root, &pattern).expect("join matches");
    assert_eq!(iter.len(), 2);
    assert!(iter.advance());
    assert!(iter.matched());
    assert!(!iter.advance());
    assert!(!iter.matched());
    assert!(iter.reset());
    assert!(iter.matched());
}

#[test]
fn source_engine_filters_nested_groups() {
    let mut root = Group::new(EngineType::TiDb);
    let mut tikv = group(EngineType::TiKv, &[Operand::Limit, Operand::Projection]);
    let mut tiflash = group(EngineType::TiFlash, &[Operand::Limit, Operand::Projection]);
    tikv.insert(GroupExpression::new(Operand::Selection));
    tiflash.insert(GroupExpression::new(Operand::Selection));
    root.insert(GroupExpression::new(Operand::TiKvSingleGather).with_child(tikv));
    root.insert(GroupExpression::new(Operand::TiKvSingleGather).with_child(tiflash));

    let pattern = tidb_planner::pattern::build_pattern(
        Operand::TiKvSingleGather,
        EngineTypeSet::TIDB_ONLY,
        [new_pattern(Operand::Limit, EngineTypeSet::TIKV_ONLY)],
    );
    let iter = new_expr_iter_from_group(&root, &pattern).expect("TiKV match");
    assert_eq!(iter.len(), 1);
}
